opendkim/opendkim-genkey.8
.TH opendkim-genkey 8 "The OpenDKIM Project"
.SH NAME
.B opendkim-genkey
- DKIM filter key generation tool
.SH SYNOPSIS
.B opendkim-genkey
[options]
.SH DESCRIPTION
.B opendkim-genkey
generates (1) a private key for signing messages using
.I opendkim(8)
and (2) a DNS TXT record suitable for inclusion in a zone file which
publishes the matching public key for use by remote DKIM verifiers.

The filenames of these are based on the selector (see below); the private
key will have a suffix of ".private" and the TXT record will have a suffix
of ".txt".
.SH OPTIONS
.TP
.I -b bits
Specifies the size of the key, in
.I bits,
to be generated.  The default is 1024 which is the value recommended by
the DKIM specification.

.TP
.I -d domain
Names the
.I domain
which will use this key for signing.  Currently only used in a comment in
the TXT record file.  The default is "example.com".

.TP
.I -D directory
Instructs the tool to change to the named
.I directory
prior to creating files.  By default the current directory is used.

.TP
.I -g granularity
Defines the key
.I granularity,
i.e. the user(s) who may use the key.  The default is "*" meaning any
user can use the key.

.TP
.I -h algorithms
Specifies a list of hash
.I algorithms
which can be used with this key.  By default all hash algorithms are allowed.

.TP
.I -n note
Includes arbitrary
.I note
text in the key record.  By default, no such text is included.

.TP
.I -r
Restricts the key for use in e-mail signing only.  The default is to allow
the key to be used for any service.

.TP
.I -s selector
Specifies the
.I selector,
or name, of the key pair generated.  The default is "default".

.TP
.I -S
Disallows subdomain signing by this key.  By default the key record will be
generated such that verifiers are told subdomain signing is permitted.

.TP
.I -t
Indicates the generated key record should be tagged such that verifiers are
aware DKIM is in test at the signing domain.
.SH NOTES
Requires that the
.I openssl(8)
binary be installed and in the executing shell's search path.
.SH VERSION
This man page covers the version of
.I opendkim-genkey
that shipped with version 1.2.0 of
.I OpenDKIM.
.SH COPYRIGHT
Copyright (c) 2007, 2008 Sendmail, Inc. and its suppliers.  All rights
reserved.

Copyright (c) 2009, The OpenDKIM Project.  All rights reserved.
.SH SEE ALSO
.I opendkim(8),
.I openssl(8)
.P
RFC4871 - DomainKeys Identified Mail


RELEASE_NOTES
			OPENDKIM RELEASE NOTES
	$Id: RELEASE_NOTES,v 1.49.2.1 2009/11/02 23:21:29 cm-msk Exp $


This listing shows the versions of the OpenDKIM package, the date of
release, and a summary of the changes in that release.

1.2.0		2009/??/??
	Feature request #SF2873902: Overhaul the database backend code so that
		features that use external files or databases can be in any
		of several supported formats.  This will make adding new
		external data sources and formats simpler, and obviates the
		need for a bunch of individual feature requests.  This may
		have a few backward compatibility issues with respect to
		the configuration file.  Feature requested by Daniel Black.
	Feature request #SF2873900: Add optional support for OpenDBX for
		connecting to ODBC and SQL backend databases.  Requested
		by Daniel Black.
	LIBOPENDKIM: Move VBR functions from libopendkim into their own
		new library, libvbr.
	LIBOPENDKIM: Add _FFR_RESIGN which allows binding of a signing handle
		to a verifying handle so that only one set of hashes needs
		to be run when a message will be re-signed as-is.  Suggested
		by Daniel Black.
	BUILD: Split up library assignments between libopendkim and opendkim,
		taking advantage of libtool.  Based on a patch by Daniel Black.

1.1.2		2009/11/01
	Under _FFR_SENDER_MACRO, need to check the value of "SenderMacro"
		in the configuration file.  Noted by Daniel Black.
	Feature request #SF2873901: Add _FFR_REDIRECT which optionally
		redirects messages that fail verification to a specific
		address, storing the original recipients in
		X-Original-Recipient: header fields.  Suggested by
		Daniel Black.
	LIBOPENDKIM: Have dkim_dns_set_callback() return
		DKIM_STAT_NOTIMPLEMENT if the underlying resolver doesn't
		have a callback facility.  Suggested by Daniel Black.
	LIBOPENDKIM: Move internal-only types and macros from dkim.h
		to dkim-internal.h.  Based on an idea from Daniel Black.
	LIBOPENDKIM: Add in all previously optional functions so that the
		API is invariant regardless of selected features.  Add
		dkim_libfeature() as a way to detect availability of features
		at runtime.  Suggested by Daniel Black.
	LIBOPENDKIM: Some global namespace consolidation.  Suggested by
		Daniel Black.
	BUILD: Fix bug #SF2882206, patch #SF2880986: Handle libtre
		installations where "--enable-system-abi" was selected.
		Reported by Stevan Bajic; patch from Daniel Black.
	BUILD: Convert libar compilation to the libtool method.
		Suggested by Daniel Black.
	BUILD: Minor autoconf fixes, contributed by Daniel Black.

1.1.1		2009/10/09
	Plug a number of potential but minor memory and file handle leaks,
		remove some dead code, guard against NULL dereferences, fix an
		errant return code check, fix a double-free(), and several
		other fixes found by a code analysis tool, used by courtesy
		of Cloudmark.
	Change "x-dkim-adsp" to "dkim-adsp" in Authentication-Results header
		field code now that RFC5617 is published.
	Apply "On-DNSError" setting when policy queries fail.
	BUILD: Some compilation fixes and type cleanup.  Based on patches
		provided by Daniel Black.
	Add _FFR_SENDER_MACRO: Determine the message sender based on the
		contents of a macro instead of on a header field.
		Based on a patch from Ondrej Sury.

1.1.0		2009/09/16
	Feature request #SF2839110: Add _FFR_IDENTITY_HEADER, to set an
       		identity (i=) for signing based on the value found in a
		particular header.  Requested by Florian Sager. 
	Fix inside dkimf_libstatus() to return extended status code 
		4.7.5 if temp-fail is due to key retrieval failure.
	Fix bug #SF2831720: Use new dynamic signature header generation
		code (see below).
	Add new exception handling code "keynotfound" (short form "key")
		and "On-KeyNotFound", which handles keys not found in DNS
		separately from other DNS errors.
	Fixes inside dkimf_libstatus() so that SMTP reply codes get set
		properly for temp-fails.
	Remove _FFR_COMMAIZE.
	LIBOPENDKIM: Add dkim_getsighdr_d(), a dynamic-length version of
		dkim_getsighdr().  The old function interface remains
		unchanged for backward compatibility.
	LIBOPENDKIM: Add dkim_dstring_printf().
	LIBOPENDKIM: Fix bug #SF2839858: Change "default_senderheaders" list
		to include only "from", per RFC5617.  Problem noted
		by Erik Lotspeich.
	Activate _FFR_SENDER_HEADERS, adding "SenderHeaders" to the
		configuration file.
	BUILD: Fix bug #SF2841499: Fix building of opendkim tools so that
		they link against libdb when necessary.
	BUILD: Fix building of libopendkim/t-test115 so it agrees with
		what SHA methods are found by libopendkim itself.
	BUILD: Don't bother building libar unless --enable-arlib is specified.

1.0.0		2009/08/14
	Initial release after code fork from dkim-milter package (v2.8.3).
	Fix bug #SF2813077: Don't do anything other than deliver messages
		when "t=y" is present in the verifying key record.  Problem
		noted by Jens Elkner.
	Fix bug #SF2835208: Set the signal mask earlier so that signals
		of interest aren't delivered to threads that can't handle
		them.  Problem noted by Mike Markley.
	LIBOPENDKIM: Fix bug #SF2795523: Correct canonicalization problem
		when a CRLF is split across body chunks and consecutive
		blank line counting is in progress.  Reported by Mark
		Martinec.
	TESTS: Fix bug #SF2813058: Fix t-test117 so it doesn't fail if
		the DKIM_TMPDIR environment variable is set.  Problem
		noted by Jens Elkner.
	PORTABILITY: Fix bug #SF2813058: Fixes to t-test100 for
		Solaris 10 (64-bit).  Patch from Jens Elkner.


opendkim/opendkim-genkey.sh
#!/bin/sh
##
## $Id: opendkim-genkey.sh,v 1.1 2009/07/16 20:51:51 cm-msk Exp $
##
## Copyright (c) 2005-2008 Sendmail, Inc. and its suppliers.
## All rights reserved.
##
## Copyright (c) 2009, The OpenDKIM Project.  All rights reserved.
##
## opendkim-genkey -- generate a key and/or TXT record for DKIM service
##
## Usage: opendkim-genkey [options]
## 	-b bits		key size (default = 1024)
##  	-d domain	signing domain (default = "example.com")
## 	-D dir		output directory (default = ".")
## 	-g gran		key granularity (default = "*")
## 	-h algs		list of allowed hash algorithms (default = unspecified)
## 	-n text		include "text" as a note in the output key record
## 	-r		restrict this key to e-mail use only
## 	-s name		selector name (default = "default")
## 	-S		disallow subdomain signing with this key
##  	-t		set the "test mode" bit in the TXT record
##
## The output of this script includes two files based on the
## selector name:
##
## 	<selector>.private	PEM-formatted private key, for use by
## 				opendkim when signing messages
##
## 	<selector>.txt		A DNS TXT record suitable for insertion
## 				into or inclusion by a DNS zone file
## 				in order to publish the public key
## 				for retrieval by verifiers


## Set up defaults
bits="1024"
domain="example.com"
outdir="."
nosubdomains=0
hashalgs=""
keygran="*"
note=""
flags=""
selector="default"
restrict=0
testmode=0
verbose=0

progname=`basename $0`

## Argument processing
while [ $# -gt 0 ]
do
	case $1 in
	-b)	if [ $# -eq 1 ]
		then
			echo $progname: -b requires a value
			exit 1
		fi

		shift
		bits=$1
		;;

	-d)	if [ $# -eq 1 ]
		then
			echo $progname: -d requires a value
			exit 1
		fi

		shift
		domain=$1
		;;

	-D)	if [ $# -eq 1 ]
		then
			echo $progname: -D requires a value
			exit 1
		fi

		shift
		outdir=$1
		;;

	-g)	if [ $# -eq 1 ]
		then
			echo $progname: -g requires a value
			exit 1
		fi

		shift
		keygran=$1
		;;

	-h)	if [ $# -eq 1 ]
		then
			echo $progname: -h requires a value
			exit 1
		fi

		shift
		hashalgs=$1
		;;

	-n)	if [ $# -eq 1 ]
		then
			echo $progname: -h requires a value
			exit 1
		fi

		shift
		note=$1
		;;

	-r)	restrict=1
		;;

	-s)	if [ $# -eq 1 ]
		then
			echo $progname: -h requires a value
			exit 1
		fi

		shift
		selector=$1
		;;

	-S)	nosubdomains=1
		;;

	-t)	testmode=1
		;;

	-v)	verbose=1
		;;

	*)	echo $progname: unknown flag $1
		exit 1
		;;
	esac

	shift
done

## do this securely and in the right place
cd $outdir
umask 077

## generate a private key
if [ $verbose -eq 1 ]
then
	echo $progname: generating private key
fi

openssl genrsa -out ${selector}.private $bits > /dev/null 2>&1

if [ $verbose -eq 1 ]
then
	echo $progname: private key written to ${selector}.private
fi

## generate a public key based on the private key
if [ $verbose -eq 1 ]
then
	echo $progname: extracting public key
fi

openssl rsa -in ${selector}.private -pubout -out ${selector}.public -outform PEM > /dev/null 2>&1

keydata=`grep -v '^-' ${selector}.public`
pubkey=`echo $keydata | sed 's/ //g'`

## output the record
if [ $testmode -eq 1 ]
then
	flags=" t=y"
fi
if [ $nosubdomains -eq 1 ]
then
	if [ "$flags" != "" ]
	then
		flags=${flags}:s
	else
		flags=" t=s"
	fi
fi

flagsout=""
if [ "$flags" != "" ]
then
	flagsout="$flags;"
fi

hashout=""
if [ "$hashalgs" != "" ]
then
	hashout=" h=$hashalgs;"
fi

noteout=""
if [ "$note" != "" ]
then
	noteout=" n=\\\"$note\\\";"
fi

if [ "$domain" != "" ]
then
	comment=" ; ----- DKIM $selector for $domain"
else
	comment=""
fi

echo ${selector}._domainkey   IN   TXT  '"'v=DKIM1\;$noteout$hashout g=$keygran\; k=rsa\;$flagsout p=$pubkey'"' $comment > ${selector}.txt

if [ $verbose -eq 1 ]
then
	echo $progname: DNS TXT record written to ${selector}.txt
fi

## all done!
rm -f ${selector}.public
exit 0
